CONTRIBUTING.md: Schema development process, improved mermaid diagram, etc.

handrews opened the pull request:
More work towards #3715 and #3716.
Also paging @jeremyfiel as GitHub won't let me add you as a reviewer.

jeremyfiel commented on CONTRIBUTING.md:
```diff
@@ -30,10 +24,20 @@ This changed in 2024, as the markdown files on `main` do not include certain cre
+### Using forks
+
 All work **MUST be done on a fork**, using a branch from the _earliest relevant and [active](#active-branches)_ `X.Y-dev` branch, and then submitted as a PR to that `X.Y-dev` branch.
 For example, if a change in November 2024 apples to both 3.1 and 3.2, the PR would go to the `v3.1-dev` branch, which will be merged up to `v3.2-dev` before the next 3.2 release.
```
```suggestion
For example, if a change in November 2024 applies to both 3.1 and 3.2, the PR would go to the `v3.1-dev` branch, which will be merged up to `v3.2-dev` before the next 3.2 release.
```

handrews replied:
🍎 🍏

jeremyfiel commented on CONTRIBUTING.md:
```diff
@@ -30,10 +24,20 @@ This changed in 2024, as the markdown files on `main` do not include certain cre
-Releases are published to the [spec site](https://spec.opanapis.org) by creating an `X.Y.Z-rel` branch where `src/oas.md` is renamed to the appropriate `versions/X.Y.Z.md` file and them merged to `main`.  The HTML versions of the OAS are automatically generated from the `versions` directory on `main`.  This renaming on the `X.Y.Z-rel` branch preserves the commit history for the published file on  ...
+### Publishing
+
+The specification and schemas are published to the [spec site](https://spec.opanapis.org) by creating an `X.Y.Z-rel` branch where `src/oas.md` is renamed to the appropriate `versions/X.Y.Z.md` file and them merged to `main`.  The HTML versions of the OAS are automatically generated from the `versions` directory on `main`.  This renaming on the `X.Y.Z-rel` branch preserves the commit history for t ...
```
```suggestion
The specification and schemas are published to the [spec site](https://spec.opanapis.org) by creating an `X.Y.Z-rel` branch where `src/oas.md` is renamed to the appropriate `versions/X.Y.Z.md` file and then merged to `main`.  The HTML versions of the OAS are automatically generated from the `versions` directory on `main`.  This renaming on the `X.Y.Z-rel` branch preserves the commit history for the published file on `main` when using `git log --follow` (as is the case for all older published files).
```

jeremyfiel reviewed (approved):
Clear and concise.  Well done!

ralfhandl commented on CONTRIBUTING.md:
```diff
@@ -57,23 +61,50 @@ Initiating the next minor release after releasing `X.Y.0`:
 Other notes:
 
-* Patch releases are _not_ merged to `dev`
+* Patch releases are _only_ merged to `dev` if they are part of the most recent release line (currently 3.1, which will shift to 3.2 once 3.2.0 is released).
```
This way `src/oas.md` will be the overall latest published version, which is nice.

However, the `versions` folder will be incomplete and only contain a subset of the files available in `main`.

This feels wrong, but I can't put my finger on what exactly bothers me.

handrews replied:
@ralfhandl that feels wrong because it's not what is happening. 🙂  I'll clarify.

ralfhandl reviewed (commented):
+1, minor nits

handrews commented on CONTRIBUTING.md:
```diff
@@ -30,10 +24,20 @@ This changed in 2024, as the markdown files on `main` do not include certain cre
+
+The specification and schemas are published to the [spec site](https://spec.opanapis.org) by creating an `X.Y.Z-rel` branch where `src/oas.md` is renamed to the appropriate `versions/X.Y.Z.md` file and them merged to `main`.  The HTML versions of the OAS are automatically generated from the `versions` directory on `main`.  This renaming on the `X.Y.Z-rel` branch preserves the commit history for t ...
+
+The publishing process for schemas is still under discussion (see issues [#3715](https://github.com/OAI/OpenAPI-Specification/issues/3715) and [#3716](https://github.com/OAI/OpenAPI-Specification/issues/3716)), with the current proposal being to release them directly from the `X.Y-dev` branch without merging to `main`, as the schemas in source control have placeholder identifiers and are not inte ...
+
```
These are not saying the same thing.  I'll see if I can figure out how to get what I mean across better.

handrews commented:
@ralfhandl I have made some changes and updated.  Changes include:

* Consistently showing the `v` prefix on the branches (apparently I couldn't make up my mind on this, but the diagram matched the branches I created so "with `v`" it is)
* Split "Development" and "Publishing" into separate sections at the same level, which hopefully makes it clear that schema development is resolved, but schema publishing (to the best of my knowledge) is not, and should be addressed in a separate PR rather than blocking this one
* Reworked the branching / merging steps to streamline them and include examples of some of the more hard-to-follow steps (this should make it clear that the "_only_" in the bit about merging patch releases to `dev` meant "only merge to `dev` if these conditions are met" and not "merge onlyt to `dev`, not `main`", as I just now figured out that was probably the source of confusion)

jeremyfiel commented on CONTRIBUTING.md:
```diff
@@ -14,66 +14,93 @@ If in doubt about a policy, please [ask on our Slack](https://communityinviter.c
+## Publishing
 
-Initial steps:
+The specification and schemas are published to the [spec site](https://spec.openapis.org) by creating an `vX.Y.Z-rel` branch where `src/oas.md` is renamed to the appropriate `versions/X.Y.Z.md` file and them merged to `main`.  The HTML versions of the OAS are automatically generated from the `versions` directory on `main`.  This renaming on the `vX.Y.Z-rel` branch preserves the commit history for ...
```
```suggestion
The specification and schemas are published to the [spec site](https://spec.openapis.org) by creating an `vX.Y.Z-rel` branch where `src/oas.md` is renamed to the appropriate `versions/X.Y.Z.md` file and then merged to `main`.  The HTML versions of the OAS are automatically generated from the `versions` directory on `main`.  This renaming on the `vX.Y.Z-rel` branch preserves the commit history for the published file on `main` when using `git log --follow` (as is the case for all older published files).
```

ralfhandl commented on CONTRIBUTING.md:
```diff
@@ -20,49 +20,45 @@ The [spec site](https://spec.openapis.org) is the source of truth.
-* Each `X.Y.Z-rel` branches from the corresponding `X.Y-dev`
-* After renaming `src/oas.md`, `X.Y.Z-rel`  merges to `main`
+* Pre-release steps:
+    * The most recent _published_ patch release from the previoius line is merged up to `vX.Y-dev`, if relevant
```
```suggestion
    * The most recent _published_ patch release from the previous line is merged up to `vX.Y-dev`, if relevant
```

ralfhandl commented:
@handrews Why do we need the `dev` branch?

If we would branch `vX.Y+1` from the same commit in `vX.Y` where `vX.Y.0-rel` was branched off, we would get the same commit history, and we would save the effort of updating the `dev` branch from two release lines. Changes would flow from the "current minor" branch to the right into the "next minor" branch, and only from `*-rel` branches to the left into `main`.

What would we loose with this simplification?

lornajane commented:
We needed the `dev` branch when switching branching setups to give us a place to rename the files before branching. I believe that at this point, the branching process is very clearly laid out. Let's follow it as it is and then see what amendments or optimisations we might use.

ralfhandl commented:
> We needed the `dev` branch when switching branching setups to give us a place to rename the files before branching.

That is my impression: the `dev` branch was temporarily necessary and can be removed from the picture without loosing anything, and significantly simplifying the picture.

handrews commented:
@ralfhandl we have [had this discussion before](https://github.com/OAI/OpenAPI-Specification/issues/3677#issuecomment-2429319683).

I just added a section explaining further.  I don't like just stacking `vX.Y-dev` branches on each other endlessly.  `dev` essentially replaces `main` for anything that doesn't need to be fully publicly published.

(BTW, we _could_ dispense with the `vX.Y.Z-rel` branches, but it would involve doing weird things, namely doing a `git mv` and then reverting half of it, and I'd rather just use git normally as much as possible.  Having that `vX.Y-Z-rel` rename commit that is _not_ on `vX.Y-dev` also gives us a better place for any future "wtf wy did the IETF move thier RFCs again" link fixes that we might need to do one day.)

lornajane reviewed (approved):
Thank you @handrews!
